### PR TITLE
Added Leaflet 1.x compatiblity

### PR DESCRIPTION
**Pull request by braandl**
Added Leaflet 1.x compatiblity.
Updated Debug / Demo Section.
Updated Dependencies.
Dropped Support for Leaflet 0.x.

### CODE REVIEW

File: src/kothic.js [hunk 1]
```diff
@@ -7,7 +7,6 @@
 var Kothic = {
 
     render: function (canvas, data, zoom, options) {
-
```
Comment (DerDakon) [inline src/kothic.js]: Needless whitespace change.

File: package.json [hunk 2]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "kothic-js",
-  "version": "0.2.0",
+  "version": "0.2.1",
```
Comment (DerDakon) [inline package.json]: My taste would be 0.3.0 because it's quite a major new thing.

File: dist/kothic-leaflet.js [hunk 3]
```diff
@@ -135,25 +160,28 @@ L.TileLayer.Kothic = L.TileLayer.Canvas.extend({
...
             if (xhr.readyState == XMLHttpRequest.DONE) {
                 if (xhr.status == 200) {
-                    window.onKothicDataResponse(JSON.parse(xhr.responseText), zoom, x, y);
+                    window.onKothicDataResponse(JSON.parse(xhr.responseText), zoom, x, y, done);
                 } else {
                     console.debug("failed:", url, xhr.status);
                 }
             }
-        }
+        };
```
Comment (DerDakon) [inline dist/kothic-leaflet.js]: Is this change really needed?

File: debug/index.html [hunk 4]
```diff
@@ -42,113 +42,116 @@
 		<div id="trace">Rendering...</div>
 	</div>
 	<script type="text/javascript">
-        var tilesUrl = 'http://osmosnimki.ru/vtile/{z}/{x}/{y}.js';
+        $( document ).ready(function() {
+
+            //var tilesUrl = 'http://osmosnimki.ru/vtile/{z}/{x}/{y}.js';
+            var tilesUrl = 'test-tile.js?&x={x}&y={y}&z={z}';
```
Comment (DerDakon) [inline debug/index.html]: Should this change persist? Looks like debug code. If yes, please just remove the old code.

File: debug/index.html [hunk 5]
```diff
@@ -42,113 +42,116 @@
...
+            var mapnik = new L.TileLayer('https://a.tile.openstreetmap.de/tiles/osmde/{z}/{x}/{y}.png', {
```
Comment (DerDakon) [inline debug/index.html]: This should probably be ${s}.tile…

File: debug/index.html [hunk 6]
```diff
@@ -42,113 +42,116 @@
...
+                maxZoom: 22
+            });
+
+            var kothic = new L.TileLayer.Kothic(tilesUrl, {
+                minZoom: 8,
+                styles: ['osmosnimki']
+            });
+
+            var map = new L.Map('map', {
+                center: new L.LatLng(49.4987159, 8.4685818), // Minsk
```
Comment (DerDakon) [inline debug/index.html]: I doubt this is still Minsk.

File: debug/index.html [hunk 7]
```diff
@@ -42,113 +42,116 @@
...
-			} else {
-                kothic.disableStyle('surface');
             }
-        });
 
+            var count = 0;
+            MapCSS.onImagesLoad = function () {
+                count += 1;
+                if (count >= 2) {
+                    console.log("adding layer kothic");
```
Comment (DerDakon) [inline debug/index.html]: forgotten debug code.

File: debug/index.html [hunk 8]
```diff
@@ -42,113 +42,116 @@
...
+                        break;
+                    case 'tiles':
+                        //kothic.disableStyle('mapnik');
+                        //kothic.disableStyle('osmosnimki');
+                        map.addLayer(mapnik);
+                        break;
+                }
+            });
+            $("#overlay").change(function () {
+                if ($("#overlay").is(':checked')) {
```
Comment (DerDakon) [inline debug/index.html]: I guess this is a browser compatibility thing. Which browsers have been tested? Which ones will not work anymore?

**Comment by nickw1** [comment]
Nice! Confirmed as working on Leaflet 1.4.0.
Would be nice to remove the jQuery dependency in the current kothic-leaflet.js; looks like it's needed to load JSON-P.

File: debug/index.html (hunk 4 above)
Comment (braandl) [reply debug/index.html]: I did it because the Webserver of http://osmosnimki.ru is not responding anymore. So i did add the Tile directly to have a working example again. I guess it should persist like this.

**Comment by braandl** [comment]
@nickw1 I removed jQuery dependency from the Kothic Lib. I only kept it in the index.html for the debug messages.

File: debug/index.html (hunk 7 above)
Comment (braandl) [reply debug/index.html]: Thanks! I removed it.

File: debug/index.html (hunk 8 above)
Comment (braandl) [reply debug/index.html]: Due to the question of @nickw1 I changed this to plain HTML, even tough it is not really part of the Kothic framework.

File: debug/index.html (hunk 6 above)
Comment (braandl) [reply debug/index.html]: I changed it back to minsk, sorry!

File: .gitignore [hunk 9]
```diff
@@ -1,2 +1,3 @@
 dist/kothic*.js
 node_modules
+.idea/
```
Comment (DerDakon) [inline .gitignore]: missing newline

File: Gruntfile.js [hunk 10]
```diff
@@ -7,8 +7,8 @@ module.exports = function (grunt) {
             options: {
                 strict: false,
 
-                bitwise: true,
-                curly: true,
+                bitwise: false,
```
Comment (DerDakon) [inline Gruntfile.js]: are these changes necessary for this feature?

File: debug/index.html [hunk 11]
```diff
@@ -42,113 +42,116 @@
...
+                styles: ['osmosnimki']
+            });
+
+            var map = new L.Map('map', {
+                center: new L.LatLng(53.8973689, 27.5416993), // Minsk
+                zoom: 12,
+                fadeAnimation: false
+            });
+
+			map.addLayer(mapnik);
```
Comment (DerDakon) [inline debug/index.html]: this indentation looks wrong

File: dist/kothic-leaflet.js [hunk 12]
```diff
@@ -60,18 +54,46 @@ L.TileLayer.Kothic = L.TileLayer.Canvas.extend({
...
+
+    drawTile: function(canvas, tilePoint, zoom, done) {
         var zoomOffset = this.options.zoomOffset,
             rzoom = zoom - zoomOffset,
             key = [rzoom, tilePoint.x, tilePoint.y].join('/'),
             url=this._url.replace('{x}',tilePoint.x).
                     replace('{y}',tilePoint.y).
                     replace('{z}',rzoom);
         this._canvases[key] = canvas;
-        if (url.endsWith('.json'))
```
Comment (DerDakon) [inline dist/kothic-leaflet.js]: Why is the non-JSON compat removed here?

**Comment by nickw1** [comment]
@braandl @DerDakon would be great to see this in the main kothic repo. I'm happy to work on this with @DerDakon's change suggestions if necessary.

**Comment by braandl** [comment]
@nickw1 I did some more 'cosmetics' on this for you. I removed the script compat because I believe it is bad style and it is not needed anymore in modern webapps. Also the grunt option changes are nececssary to build with newer grunt versions. If you are dedicated to enhance this even more, you might think about switching to webpack or another modern javascript stack, other than the rather outdated grunt.

**Comment by nickw1** [comment]
@braandl thanks - but just to clarify I am not the maintainer, that is @DerDakon. I am just a kothic user/very occasional contributor (have made 2 PRs). It's up to @DerDakon on whether it's good to merge now. Apologies for any confusion I may have caused!

**Comment by DerDakon** [comment]
I'm not really a maintainer, I just use to have commit access. One obvious thing is that this branch seems to need a rebase as it shows collisions on kothic.js.

**Comment by Komzpa** [comment]
To cause more confusion: I just came in here out of random reason, found an ongoing conversation and that there are lively PRs, and also that for some reason my notifications were switched off for this repo. To decrease bus factor, I've added @nickw1 to collaborators, as I know long time user of Kothic JS and thus a valuable stakeholder, you should be able to accept PRs too now.

**Comment by nickw1** [comment]
@Komzpa thanks for adding me as collaborator, @braandl I will check out your PR and verify all is working.

**Comment by nickw1** [comment]
@braandl to update you: have created a new branch leaflet1.0 in the main kothic repo and am merging your PR into that, so that any issues can be resolved before it goes in master.
